### ISSUE TITLE
operator: failing to reconcile grafana agent config Secret writing too much data

### ISSUE THREAD

**Issue by stevo-f3** [grafana/agent#1649]
With two MetricsInstance resources and quite a few Pod+ServiceMonitors on a k8s cluster, grafana agent operator will try to write to grafana agent config Secret too much data, more than can fit k8s Secret / etcd constraints.

error message:
```
failed to reconcile secret: failed to update service: Secret "grafana-agent-config" is invalid: data: Too long: must have at most 1048576 bytes
```

Context:
- grafana agent operator
  - chart 0.1.10
  - binary v0.24.1
- grafana agent: v0.24.1

Please consider removing the limitation by integrating operator and agent differently, not via Secret. E.g. have operator store config locally and expose it via API to the agent config reloader, have agent config reloader retrieve config via URL programmed to agent by operator.

**Comment by captncraig** [comment]
First off, that does seem very large for an agent config. But it could happen if you have a large number of metrics instances or scrape configs I suppose. In general, I'd recommend trying to reduce that as much as possible by consolidating scrape configs. We scrape a large number of pods with a single scrape config that grabs anything with a port named `*_metrics`, and deploy all of our workloads accordingly. If you could reduce the overall number of scrape configs, that would probably be ideal, and you would also save agent resources since it would not have to do so many independent service discovery tasks.

But if that is not feasible... we brainstormed a few different _possible_ implementations.

1. Compress config, store it in secret, and run an init container to decompress it.
  I don't love this since the secret is less useful for debugging, and it only solves the problem until the compressed size grows to over a megabyte.

2. Split the config into 750k or so "chunks", and store each in its' own secret. Mount them all to the agent, and have an init container concatenate them. 
  This would scale arbitrarily, and keep configs readable, at the cost of a little bit of complexity, and one ugly bash hack. And in the 90% case where the config is small, we don't have to change anything.

3. Some kind of shared PV setup? I think this is too dependent on the storage capabilities of the cluster to be reliable.

4. Have operator save the config and serve it at `/configs/{hash}` or something and have the agent read it as remote config. 
  This would scale to arbitrary size, and be relatively simple to implement, but it does make the agent dependent on the operator being up and running in order for it to start. That may or may not be desirable, so this would need to be an opt-in feature with the secret remaining the default I think.

I think the main contenders are numbers 2 and 4. 4 seems pretty fun to implement, but I imagine 2 may be more stable and predictable.

**Comment by stevo-f3** [comment]
@captncraig thanks for sharing! Agree, 2. and 4. options make sense.

Btw, Flux v2 uses approach similar to 4. for integrating sources (GitRepo, HelmRepo etc managed by source-controller) with helm releases (HelmRelease, HelmChart  managed by helm-controller); source has a status, and when ready it holds reference/URL to the helm package that helm-controller can retrieve.

With 4. sidecar and operator would change, while I think agent should stay unchanged. As sidecar `prometheus-config-reloader` is currently used, so if not modify it to support config from URL (besides from file), than maybe add one more side car. That sidecar can still use Secret/ConfigMap as input, with some new reserved data property (Grafana agent config URL?) which when present in Secret/ConfigMap would tell sidecar to retrieve actual config content using from URL strategy. Sidecar watches for Secret/ConfigMap and updates agent config in filesystem on Secret/ConfigMap changes (for reloader sidecar to use). Agent (and all sidecars) should be able to start with empty / not present config, so I see no special blocking on operator availability; if URL is present but isn't accessible for whatever reason, emit event on Agent CR, keep current config. Agent operator updates Secret/ConfigMap on GrafanaAgent/MetricsInstance/*Monitor/... changes. Since Agent operator is aware of the config size, it can even dynamically decide on strategy, and fallback to URL only once config size is over some threshold. All the new resources (new Service for operator? or change of existing grafana-agent-operated Service) and changes (extra sidecar, feature flag for operator binary) to StatefulSet can be behind (opt-in) feature flag.

**Comment by zytek** [comment]
Please reopen, as this issue is not fixed and we stumbled upon it on a relatively small cluster (620 pods)

**Comment by zytek** [comment]
Re config approach, you could mirror approach used in kube-prometheus-stack with sidecard reloader.

**Comment by ajgrande924** [comment]
bump, would really need this also

**Comment by rfratto** [comment]
I'll reopen because I noticed the comments, but in general we'd like for people to create new issues for anything that the stalebot (which we stopped using) closed, because we might miss the notifications.

**Comment by akasprzok** [comment]
I'm also running into this - we make heavy use of `params` in serviceMonitors, leading to rather large grafana-agent-configs.

**Comment by rfratto** [comment]
Even though #3295 isn't a permanent fix (I think sharding the secrets would be the permanent fix here), I think it's sufficient enough that we can consider this issue closed. If anyone is still experiencing this problem despite #3295, please open a new issue and we can track it there.
